nixos/modules/services/cluster/k3s/default.nix

```
@@ -78,7 +78,8 @@ in
 
     systemd.services.k3s = {
       description = "k3s service";
-      after = mkIf cfg.docker [ "docker.service" ];
+      after = [ "network-online.service" "firewall.service" ] ++ (optional cfg.docker "docker.service");
+      wants = [ "network-online.service" "firewall.service" ];
```

k3s seems robust enough to work without ip addresses.
 This also matches how configure docker: https://github.com/NixOS/nixpkgs/pull/117618/files

```suggestion
      after = [ "network.service" "firewall.service" ] ++ (optional cfg.docker "docker.service");
      wants = [ "network.service" "firewall.service" ];
```